llvm/lib/CodeGen/GlobalMerge.cpp

```diff
@@ -632,10 +633,13 @@ void GlobalMergeImpl::setMustKeepGlobalVariables(Module &M) {
   for (Function &F : M) {
     for (BasicBlock &BB : F) {
       Instruction *Pad = BB.getFirstNonPHI();
-      if (!Pad->isEHPad())
+      auto *II = dyn_cast<IntrinsicInst>(Pad);
+      if (!Pad->isEHPad() &&
+          !(II && II->getIntrinsicID() == Intrinsic::eh_typeid_for))
         continue;
```

Comment:
```suggestion
      if (!Pad->isEHPad())
        continue;
        
      if (auto *II = dyn_cast<IntrinsicInst>(Pad))
        if (II->getIntrinsicID() == Intrinsic::eh_typeid_for))
          continue;
```
nit: I'd separate the two ifs here.

Reply:
We can't split the if statements this way since we want to continue if its neither an EHPad instruction or an Intrinsic::eh_typeid_for instruction.

Alternate could be:
```
      if (!Pad->isEHPad())
        if (auto *II = dyn_cast<IntrinsicInst>(Pad))
          if (II->getIntrinsicID() != Intrinsic::eh_typeid_for)
            continue;
```

Reply:
Oh sorry, I completely misread the code here. I think your original version was already the best one then.